File: clients/ios/StellarChat/StellarChat/Models/VideoCache.swift
import Foundation

/// Disk-only encrypted cache for decrypted video files.
/// Key = blobHash (SHA-256 hex of encrypted blob).
/// No in-memory cache — videos are too large for NSCache.
final class VideoCache {
    static let shared = VideoCache()

    private let cacheDir: URL
    private let tempDir: URL

    private init() {
        let caches = FileManager.default.urls(
            for: .cachesDirectory, in: .userDomainMask
        )[0]
        cacheDir = caches.appendingPathComponent("StellarVideoCache", isDirectory: true)
        tempDir = caches.appendingPathComponent("StellarVideoTemp", isDirectory: true)
        try? FileManager.default.createDirectory(at: cacheDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
    }

    /// Check if a video is cached. Returns a decrypted temp file URL for playback, or nil.
    func cachedFileURL(for blobHash: String) -> URL? {
        let encFile = cacheDir.appendingPathComponent(blobHash)
        guard FileManager.default.fileExists(atPath: encFile.path) else { return nil }
        // Decrypt to a temp file for playback
        guard let encrypted = try? Data(contentsOf: encFile),
              let decrypted = try? StorageEncryption.decrypt(encrypted) else { return nil }
        let playbackURL = tempDir.appendingPathComponent(blobHash).appendingPathExtension("mp4")
        try? decrypted.write(to: playbackURL)
        return playbackURL
    }

    /// Store decrypted video data to encrypted disk cache. Returns a temp file URL for immediate playback.
    func store(_ videoData: Data, for blobHash: String) -> URL {
        // Encrypt and save to disk cache
        if let encrypted = try? StorageEncryption.encrypt(videoData) {
            let encFile = cacheDir.appendingPathComponent(blobHash)
            try? encrypted.write(to: encFile)
        }
        // Write decrypted to temp for immediate playback
        let playbackURL = tempDir.appendingPathComponent(blobHash).appendingPathExtension("mp4")
        try? videoData.write(to: playbackURL)
        return playbackURL
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/blossom/VideoCache.kt
package chat.onym.android.blossom

import android.content.Context
import chat.onym.android.crypto.StorageEncryption
import java.io.File

/**
 * Disk-only encrypted cache for decrypted video files.
 * Key = blobHash (SHA-256 hex of encrypted blob).
 * No in-memory cache — videos are too large for LruCache.
 */
class VideoCache private constructor(context: Context) {
    private val cacheDir: File = File(context.cacheDir, "StellarVideoCache").also { it.mkdirs() }
    private val tempDir: File = File(context.cacheDir, "StellarVideoTemp").also { it.mkdirs() }

    /** Check if a video is cached. Returns a decrypted temp file for playback, or null. */
    fun getCachedFile(blobHash: String): File? {
        val encFile = File(cacheDir, blobHash)
        if (!encFile.exists()) return null
        return try {
            val encrypted = encFile.readBytes()
            val decrypted = StorageEncryption.decrypt(encrypted)
            val playbackFile = File(tempDir, "$blobHash.mp4")
            playbackFile.writeBytes(decrypted)
            playbackFile
        } catch (_: Exception) {
            null
        }
    }

    /** Store decrypted video data to encrypted disk cache. Returns a temp file for immediate playback. */
    fun store(blobHash: String, videoData: ByteArray): File {
        try {
            val encrypted = StorageEncryption.encrypt(videoData)
            File(cacheDir, blobHash).writeBytes(encrypted)
        } catch (_: Exception) { }
        val playbackFile = File(tempDir, "$blobHash.mp4")
        playbackFile.writeBytes(videoData)
        return playbackFile
    }

    companion object {
        @Volatile
        private var instance: VideoCache? = null

        fun getInstance(context: Context): VideoCache {
            return instance ?: synchronized(this) {
                instance ?: VideoCache(context.applicationContext).also { instance = it }
            }
        }
    }
}
